### ISSUE TITLE
MimeTypeUtils performance degradation

### ISSUE THREAD

**Issue by Percy0601** [spring-projects/spring-framework#24671]
85.32%(255967.32 ms) org.springframework.util.MimeTypeUtils$ConcurrentLruCache.get
85.32%(255967.32 ms) org.springframework.util.MimeTypeUtils.parseMimeType

<img width="1211" alt="DX-20200310@2x" src="https://note.youdao.com/yws/api/personal/file/WEB0f3570d56d4a6d440513e0bddef33524?method=download&shareKey=44e66b49d419cb1ae67a54e99c851b16">

https://stackoverflow.com/questions/59879550/spring-boot-2-2-x-increased-cpu

It's similar to what this link describes, but I Running 10 days later Appear

spring boot 2.1.x is well, but 2.2.x cpu 90%

**Comment by bclozel** [comment]
A few queries per-second is not in line with recent issues and the benchmarks we've run, so there might be another problem here.

The screenshot you've provided point to content types produced by the application itself, so you should be able to know which media types are produced here, even without specific monitoring (looking at the controllers and the web configuration for those endpoints might be enough).

Does this happen on all services running on this Spring Boot version, or on a particular application? Anything you can give us here will help us find the issue.

**Comment by Percy0601** [comment]
My workmate service has the similar problem when the concurrency is relatively large. The service is about 500qps, but after removing the **Actuator** module and **ErrormvcAutoConfiguration**, back to normal.

`
  servlet:
    multipart:
      enabled: false
  autoconfigure:
    exclude: org.springframework.boot.autoconfigure.web.servlet.error.ErrorMvcAutoConfiguration

management:
  endpoints:
    enabled-by-default: false
    web:
      base-path: ""
`

OR remote actuator starter

Thanks.

**Comment by bclozel** [comment]
Excluding the error page auto-configuration and removing actuator are very different things.
Removing the error auto-configuration means you won't have custom error pages (and Spring Boot won't render JSON error pages, for example). On the other hand, removing the Actuator dependency will remove micrometer instrumentation and actuator endpoints; unless there's heavy traffic on actuator endpoints, this does not explain the performance difference.

I'm reopening this issue to find the actual source of the performance problem.
You've described very different scenarios and use cases.

Could you narrow down the problem and maybe provide a sample? I'll try and run comparative benchmarks on my side to try and pinpoint possible issues.

**Comment by rstoyanchev** [comment]
Note related #24469, perhaps a closer look there could also provide clues here?

**Comment by Percy0601** [comment]
When Mime Type.queue is more than half of the default, (the default value is 64, which is more than 32), the code not only performs read lock, but also write lock.
Although the LRU algorithm is subsequently used, the program will not be automatically reduced to less than half.

**Comment by bclozel** [comment]
@Percy0601 I'm sure that I can find a performance difference with this change.
Your first use case (long running, low request/sec) might explain that after many hours, we might exceed the 32 limit. But the low request/sec would not explain such a difference, since locks really make a difference with high concurrency.

Again, I'm looking at various benchmarks and improvements - any information about the issues you're seeing will help. Also, could you explain a bit more the screenshot you've provided? The 85% number doesn't tell much: was this profiling performed in production? How long had the application been running? What was the traffic at the time? Is that 85% the overall % of CPU time, or a subset? Is the application itself resolving media types as part as the application code?

**Comment by Percy0601** [comment]
I also need to explain another phenomenon. After removing the spring-actutor-starter related components, the number of MimeTypeUtils.get () method reduce half.
By default, an HTTP request requires 16 times to MimeTypeUtils.get (), and 8 times after remove.

**Comment by Percy0601** [comment]
85% is JVM, Not Linux System.

**Comment by Percy0601** [comment]
The program started around the 7th day of operation. On the 9th day, the response time of a code without any logic over 300ms. On the 10th day, the service was disabled by the service management mechanism, and then various alerts were issued.

**Comment by Percy0601** [comment]
In order to verify the problem today, I used more than 32 common media types for testing. The response time of a non-logical method exceeded 15ms. Before that, it took an average of 0ms or 0.1ms.

**Comment by bclozel** [comment]
Thanks @Percy0601 , this is just what I needed! I'll work on #24469 and I also have an alternate parser implementation as well. Cheers!

**Comment by Percy0601** [comment]
Because using the spring-actuator-starter component caused MimeType.get () calls to increase twice times, this issue should be reported back to the spring-boot community, which is also a reason that will affect performance.

**Comment by Percy0601** [comment]
I am going to bed and hope that the issue will be fixed soon. Thanks

**Comment by bclozel** [comment]
No need to report that in Spring Boot, all the implementation is here. Having a reasonable cache size should not make the performance way worse. I've scheduled this for the next maintenance release and Spring Boot should pick this up.

Thanks a lot for your help!

**Comment by Percy0601** [comment]
Dear @bclozel , I have a question, because I read that the implementation code uses ConcurrentHashMap, so why use readLock and writeLock?
I think the main function is to improve performance, and ConcurrentHashMap cannot add the same key-value pairs repeatedly, which is also thread-safe
So my conclusion is, what is the impact of the code if locks are not used? If I did it, I might not lock it.
Ask the experts for answers, Thanks

**Comment by bclozel** [comment]
I've merged #24469 and improved the proposed changes.

I've run benchmarks with changes and they're showing a +75% throughput improvement in the following case:
* high concurrency
* cache size > 32 elements

It seems that reducing the scope of the read lock and keeping track of the size locally improves things. 

I've also removed the artificial `cacheMaxSize / 2` limit; from now, we'll only keep track of least/most used cache entries as soon as the cache hits its maximum size. This generally improves cases where the limit is not reached, at the cost of possible inefficiencies when the limit has just been reached (trashing cached result that might be needed still).

This last change is a x10 throughput improvement if the cache size is under its limit.

**Comment by hsw409328** [comment]
> When MimeTypeUtils.queue is more than half of the default, (the default value is 64, which is more than 32), the code not only performs read lock, but also write lock.
> Although the LRU algorithm is subsequently used, the program will not be automatically reduced to less than half.

This problem occurs in the formal environment. Offline test is normal. Can you tell me the stable recurrence mode?

**Comment by hsw409328** [comment]
> In order to verify the problem today, I used more than 32 common media types for testing. The response time of a non-logical method exceeded 15ms. Before that, it took an average of 0ms or 0.1ms.

你好，我正式环境出现了个问题，但是本地一直无法复现，请教一下复现的方式。我这边也使用了多种媒体类型，强制每次都触发读写锁，但是依然无法复现。

**Comment by bclozel** [comment]
@hsw409328 Could you provide more information? To which version you’re comparing to?
What kind of traffic you’re getting in that application? Is that a benchmark or real traffic?

Anything that you can provide to help narrow down the issue would be helpful.
With the changes done in this issue, the 32 limit has no effect and should behave like <32.

**Comment by hsw409328** [comment]
> @hsw409328 Could you provide more information? To which version you’re comparing to?
> What kind of traffic you’re getting in that application? Is that a benchmark or real traffic?
> 
> Anything that you can provide to help narrow down the issue would be helpful.
> With the changes done in this issue, the 32 limit has no effect and should behave like <32.

spring boot 2.2.6 
qps 400
dump a lot:
"http-nio-1601-exec-919" #1195 daemon prio=5 os_prio=0 tid=0x00007fe4c838b000 nid=0x8a3b2 waiting on condition [0x00007fdf2c583000]
   java.lang.Thread.State: WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x00000000c1c33a60> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
        at org.apache.commons.pool2.impl.LinkedBlockingDeque.takeFirst(LinkedBlockingDeque.java:590)
        at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:432)
        at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:349)


dump jmap:
 num     #instances         #bytes  class name
----------------------------------------------
   1:       4071536       97716864  java.util.concurrent.ConcurrentLinkedQueue$Node
   2:         50959       77795112  [B
   3:        136164       62161192  [C
   4:          1010        4140848  [Ljava.nio.ByteBuffer;
   5:        119791        2874984  java.lang.String
   6:         57984        2783232  org.apache.tomcat.util.buf.ByteChunk
   7:         53940        2589120  org.apache.tomcat.util.buf.CharChunk
   8:         52930        2540640  org.apache.tomcat.util.buf.MessageBytes
   9:         64236        2055552  java.util.concurrent.ConcurrentHashMap$Node
  10:         29694        1770824  [Ljava.lang.Object;
  11:         13858        1426632  java.lang.Class
  12:         15448        1359424  java.lang.reflect.Method
  13:         34318        1098176  java.util.HashMap$Node
  14:         10047        1045240  [Ljava.util.HashMap$Node;
  15:         20992        1041520  [Ljava.lang.String;
  16:         16244         909664  java.util.LinkedHashMap
  17:         12759         891640  [I
  18:           389         666200  [J
  19:         20506         656192  java.lang.ThreadLocal$ThreadLocalMap$Entry
  20:         15472         618880  java.util.LinkedHashMap$Entry
  21:          6276         589024  [Z
  22:          1291         565200  [Ljava.util.concurrent.ConcurrentHashMap$Node;
  23:         11671         560208  java.nio.HeapByteBuffer
  24:         20986         503664  java.util.ArrayList
  25:         10087         484176  java.util.HashMap
  26:          3068         441792  java.text.DecimalFormat
  27:         26221         419536  java.lang.Object

**Comment by hsw409328** [comment]
@bclozel 
![test](https://user-images.githubusercontent.com/11609045/78752517-d5fed080-79a6-11ea-9695-c983d7178c30.jpg)

**Comment by bclozel** [comment]
@Percy0601 @hsw409328 I'm trying to find the change that caused the performance issue and how to reproduce that properly in a benchmark.
Could you provide some more data, comparing the version that was performing well (Spring Framework version, ideally flamegraph or any performance metric) and the latest current version (with same flamegraph and performance metrics).
Thanks a lot for your help.

**Comment by yes-github** [comment]
我的正式环境SpringBoot2.2.x多次碰到相同的问题, 一段时间后CPU高居不下, 下面是用arthas抓到的占用CPU高的线程信息:

[arthas@41835]$ thread -b
"XNIO-1 task-33" Id=546 RUNNABLE
    at java.util.concurrent.ConcurrentLinkedQueue.remove(ConcurrentLinkedQueue.java:490)
    at org.springframework.util.MimeTypeUtils$ConcurrentLruCache.get(MimeTypeUtils.java:454)
    at org.springframework.util.MimeTypeUtils.parseMimeType(MimeTypeUtils.java:196)
    at org.springframework.http.MediaType.parseMediaType(MediaType.java:571)
    at org.springframework.http.HttpHeaders.getContentType(HttpHeaders.java:966)
    at org.springframework.http.converter.AbstractHttpMessageConverter.addDefaultHeaders(AbstractHttpMessageConverter.java:260)
    at org.springframework.http.converter.AbstractGenericHttpMessageConverter.write(AbstractGenericHttpMessageConverter.java:88)
    at org.springframework.web.servlet.mvc.method.annotation.AbstractMessageConverterMethodProcessor.writeWithMessageConverters(AbstractMessageConverterMethodProcessor.java:295)
    at org.springframework.web.servlet.mvc.method.annotation.RequestResponseBodyMethodProcessor.handleReturnValue(RequestResponseBodyMethodProcessor.java:181)
    at org.springframework.web.method.support.HandlerMethodReturnValueHandlerComposite.handleReturnValue(HandlerMethodReturnValueHandlerComposite.java:82)
    at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:124)
    at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:888)
    at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:793)
    at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
    at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040)
    at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)
    at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
    at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:523)
    at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:590)
    at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
    at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:320)
    at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:126)
    at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:90)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
    at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:118)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
    at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
    at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
    at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:158)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
    at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
    at org.springframework.security.web.authentication.www.BasicAuthenticationFilter.doFilterInternal(BasicAuthenticationFilter.java:154)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
    at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
    at org.springframework.security.web.header.HeaderWriterFilter.doHeadersAfter(HeaderWriterFilter.java:92)
    at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:77)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
    at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
    at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
    at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215)
    at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
    at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:358)
    at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:271)
    at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
    at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
    at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
    at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
    at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
    at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
    at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
    at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
    at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:108)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
    at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
    at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
    at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
    at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
    at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
    at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
    at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
    at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
    at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
    at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:68)
    at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
    at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
    at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
    at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
    at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
    at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
    at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
    at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
    at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
    at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
    at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:269)
    at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:78)
    at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:133)
    at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:130)
    at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
    at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
    at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:249)
    at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:78)
    at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:99)
    at io.undertow.server.Connectors.executeRootHandler(Connectors.java:376)
    at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)

    Number of locked synchronizers = 2
    - java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync@1266f00b <---- but blocks 96 other threads!
    - java.util.concurrent.ThreadPoolExecutor$Worker@1ff013b8

Affect(row-cnt:0) cost in 746 ms.

=========================================================

**Comment by bclozel** [comment]
I think I’ve found it, thanks for the information all!
I’ll create a new issue and link back here as it’s a different problem.

**Comment by bclozel** [comment]
See #24886
